### PR TITLE
Ensure background submit works for marketo modals

### LINKED ISSUE #9208: Background submit CUBE beta apply for access form
Submit form to marketo without being redirected to the thank-you page

### PR DESCRIPTION
**Pull request by bethcollins92**
## Done

- Fix modal marketo forms to prevent redirect, displaying a thank you message inside the modal

## QA

- Check out this feature branch
- Run the site using the command `./run serve` or `dotrun`
- View the site locally in your web browser at: http://0.0.0.0:8001/cube 
- Select the 'Apply for access' button at the bottom of the page, fill out the form and check on submit the form doesn't redirect, but instead shows a thank you message

## Issue / Card

Fixes #9208

### CODE REVIEW

**Review by anthonydillon — COMMENTED** [review]
Codewise this looks good with a simple comment

**Comment by bethcollins92** [comment]
@barrymcgee I made a couple of changes, the path for `bg-form-submit.js` in the webpack.config file was wrong so the run script was failing. Also I added the classes `js-pagination--1` and `js-pagination--2` back to the `cube-beta.html` file as without these the close modal was breaking

File: static/js/src/dynamic-forms.js [hunk 1]
```diff
@@ -24,7 +26,8 @@
       recaptchas.forEach(function (field) {
         if (!field.hasAttribute("data-widget-id")) {
           let siteKey = field.getAttribute("data-sitekey");
-          const recaptchaWidgetId = grecaptcha.render(field, {
+          const gc = grecaptcha || null;
```
[the remark this answers is not in the document]
Comment (barrymcgee) [reply static/js/src/dynamic-forms.js:29]: This is an error - this should be wrapped in an `onLoadCallback` as per https://developers.google.com/recaptcha/docs/display

**Comment by barrymcgee** [comment]
Great work!

File: static/js/src/dynamic-forms.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (barrymcgee) [reply static/js/src/dynamic-forms.js:29]: @bethcollins92 Were you able to correct this?

[the remark this answers is not in the document]
Comment (bethcollins92) [reply static/js/src/dynamic-forms.js]: I believe so, would you be able to check the code and QA it?

**Comment by barrymcgee** [comment]
@anthonydillon Digging into that `reCaptcha` loading comment, I found the root of the issue was a race condition - the Recaptcha `render` function to populate the page element was being called before the library was being initialised. 

This is because the lib was being loaded in dynamic-forms via Javascript but this was bundled into main.js and loaded on every page regardless. 

As a result, I've made a change to remove this dynamic loading and add the Recaptcha lib to the base template so it loads directly from base.html, using async and defer to mitigate against perf considerations. 

I grepped the codebase and found this lib was being loaded multiple times in various form partials. I've now removed all of those to prevent duplicate, redundant HTTP requests.

Can you sanity check please?

File: static/js/src/bg-form-submit.js [hunk 2]
```diff
@@ -0,0 +1,205 @@
...
+      }
+    });
+  }
+
+  // recaptcha test
+  const recaptchaListElement = marketoForm.getElementsByClassName(
+    "g-recaptcha"
+  )[0];
+  const recaptchaWidgetId = recaptchaListElement.dataset.widgetId;
+  const response = grecaptcha.getResponse(recaptchaWidgetId);
```
Comment (anthonydillon) [inline static/js/src/bg-form-submit.js:135]: I don't see how there could be a race condition on this as this function is only called when the form is submitted. I think we should leave the recapture loading per form and dynamically for interactive forms.

**Comment by anthonydillon** [comment]
Removing the dynamic loading of recaptcha for displaying an interactive form brakes the interactive forms as the recaptcha needs to be initialised after the form is rendered.

File: static/js/src/bg-form-submit.js (hunk 2 above)
Comment (barrymcgee) [reply static/js/src/bg-form-submit.js:135]: Ok, reverted.

**Comment by bethcollins92** [comment]
QA'd again since reverting the code. Checked all the interactive modals, all LGTM apart from the `ReferenceError: grecaptcha is not defined`, which doesn't seem to be effecting the functionality of the form

**Comment by anthonydillon** [comment]
Thanks Beth I'm sorry working on that error.

**Comment by anthonydillon** [comment]
@bethcollins92 @barrymcgee Sorry it took so long. I've been in meetings. Found the culprit. We were manually calling the async callback in a synchronous which was before the recaptcha script had initialised.

**Comment by anthonydillon** [comment]
I have QA'ed the static and dynamic forms and have confirmed they are sending to Marketo. 

@caldav could you confirm a test submission is reaching Marketo.
@pmahnke could you QA for edge cases you are aware of please.

**Review by anthonydillon — APPROVED** [review]
Code is much clearer thanks :+1:

**Comment by barrymcgee** [comment]
> We were manually calling the async callback in a synchronous which was before the recaptcha script had initialised.

I knew it was a race condition somewhere - thanks for digging in

**Comment by pmahnke** [comment]
Seems to work well and data gets to Marketo.

+1

- http://0.0.0.0:8001/openstack#get-in-touch - good
- http://0.0.0.0:8001/download/server# - good
- http://0.0.0.0:8001/engage/enterprise-kubernetes-guide/thank-you - good, not a modal
- http://0.0.0.0:8001/openstack/tco-calculator#get-in-touch - good
